### ISSUE TITLE
git bad signature index file corrupt

### ISSUE THREAD

**Issue by giuse88** [tpope/vim-fugitive#648]
Hi, 

Often vim-fugitive corrupts my git index. This happens when I call `Gstatus` and then I add the same file more the one with `-`.  Any one else has this problem?

**Comment by tpope** [comment]
Next time it happens take a look inside the index file and see if the contents of the `:Gstatus` window have overwritten it.

**Comment by cj** [comment]
@giuse88 @tpope not sure if this is related, but on the latest version of vim-fugative after I `wq` to save a commit I get this error: ![](http://i.imgur.com/OqO9Px0.png) and then this shows up ![](http://i.imgur.com/kRsz1x3.png).

If I use vim to check `.git/index` I see: ![](http://i.imgur.com/KRT4OGt.png)
using nano I see: ![](http://i.imgur.com/mMmbqOi.png)

**Comment by tpope** [comment]
Sounds like an issue with whatever this "BufSurf" plugin is.

**Comment by cj** [comment]
this is the plugin https://github.com/ton/vim-bufsurf.  the issue only just started happening after I updated fugitive.  I guess something change that bufsurf doesn't like.

**Comment by tpope** [comment]
If you can figure out the exact commit that triggered it I can further advise.

**Comment by mataps** [comment]
i'm having the same issue when doing Gcommit then I added a file to stage. After that, I did another Gcommit, this time with a the message and I got this error.

**Comment by cj** [comment]
@tpope so I'm still having this issue.  I no longer have that plugin, but after I commit I'm still getting this:

![](http://i.imgur.com/3OJ3nAd.png)

**Comment by tpope** [comment]
Please elaborate.

**Comment by cj** [comment]
After I do a `:Gwrite` then `:Gcommit` type in my message and then `:wq` this buffer just pops up.

**Comment by tpope** [comment]
So the commit message buffer closes and then this opens in a split? That's odd. It looks like a failed attempt to open `:Gstatus`, which is what happens when you call `:Gcommit` with no changes.

Can you check the value of `:echo b:fugitive_commit_arguments` before you `:wq`?

**Comment by cj** [comment]
Strange, because there was a commit as it allowed me to enter the message before I `:wq`'d.

here's the value: `--cleanup=strip -F /home/cj/apps/acd/autolink/.git/COMMIT_EDITMSG  --no-edit --no-interactive --no-signoff`

**Comment by tpope** [comment]
While the commit message is still open, try running the following in another window:

     GIT_EDITOR=false git commit --cleanup=strip -F /home/cj/apps/acd/autolink/.git/COMMIT_EDITMSG --no-edit --no-interactive --no-signoff

Check for errors, and the exit status (`echo $?`).

**Comment by cj** [comment]
![](http://i.imgur.com/NfJH1ns.png)

**Comment by fourjay** [comment]
I've run into this (pretty consistently). I've tried the troubleshooting
steps without seeing much different than reported above. That said I
think I might have found a workaround (and thus a pointer) My (less then
full perusal) of the code shows that the commit window is normally set
with ``bufhidden=bdelete`` but on commit it's set to ``bwipeout``. It
looks like some path allows that to get skipped. After a commit I can
still see the hidden .git/index file with ls! If I switch to the file at
that point (either directly or by invoking Gstatus) I see the garbled
binary index.

To clarify a bit:

1) I use Gtatus. Everything is good
2) I commit a change with Gcommit
3) At that point I can still see the .git/index file with ls!
4) If switch to the window, it's garbled
5) If I close that window and re-invoke Gstatus, it's back to expected.

If I bwipeout the buffer directly ( ``:[BNUMBER]bwipeout`` ) then
Gstatus and cousins work fine.

The workaround (and possible clue) is that I typically use ZZ as a wq
shortcut. In after/gitcommit.vim I've now set a function + mapping that
calls bwipeout and the problem goes away (aka I remap ZZ)

Hoping this is useful :-)

**Comment by yevhen-m** [comment]
@cj I have the same issue with garbled text after `GCommit` command. Have you managed to fix it somehow?

**Comment by unChartedZone** [comment]
Is anyone here who is having this issue, using vim-airline by any chance?

**Comment by yevhen-m** [comment]
@unChartedZone I use vim-airline.

**Comment by unChartedZone** [comment]
@yevhen-m I noticed that the issue with fugitive goes away when I uninstall vim-airline, so perhaps that plugin could be the cause of the issue but I'm not entirely sure.

**Comment by pedrofranceschi** [comment]
I'm having the same issue… Any fix available?

**Comment by darkalia** [comment]
Hello, I can confirm that this is related to airline. I hit this after switching to Xenial (the exact same setup was working fine on 15.10).

Here is my .vimrc file
`
set nocompatible              " required
filetype off                  " required
syntax on

" Change leader
let mapleader="\<Space>"

" set the runtime path to include Vundle and initialize
set rtp+=~/.vim/bundle/Vundle.vim
call vundle#begin()

" alternatively, pass a path where Vundle should install plugins
"call vundle#begin('~/some/path/here')

" let Vundle manage Vundle, required
Plugin 'gmarik/Vundle.vim'

" Add all your plugins here (note older versions of Vundle used Bundle instead
" of Plugin)*

" Auto-Complete
Bundle 'Valloric/YouCompleteMe'

" Folding
Plugin 'tmhedberg/SimpylFold' "Folding

" Python autodindent
Plugin 'vim-scripts/indentpython.vim' " Indent python

" Syntax
Plugin 'scrooloose/syntastic' "Syntax checker

"
Plugin 'nvie/vim-flake8' " PEP8

" Colors
Plugin 'jnurmine/Zenburn' " Zenburn theme (console)
Plugin 'altercation/vim-colors-solarized' "Solarized theme (GUI)

Plugin 'scrooloose/nerdtree' "Hierarchy tree

Plugin 'kien/ctrlp.vim' " Search files

Plugin 'vim-airline/vim-airline' "Status bar
Plugin 'vim-airline/vim-airline-themes'

Plugin 'tpope/vim-fugitive' " Show git branch in status bar

Plugin 'xolox/vim-misc' " Needed by vim-session
Plugin 'xolox/vim-session' " Make use of mksession easier

Plugin 'saltstack/salt-vim' " For saltstack files

Plugin 'fatih/vim-go' " Go Plugin

Plugin 'pearofducks/ansible-vim' " Ansible plugin

Bundle 'Rykka/riv.vim'

" All of your Plugins must be added before the following line
call vundle#end()            " required
filetype plugin indent on    " required
set encoding=utf-8

" Set filetype based on name
autocmd BufRead,BufNewFile *Dockerfile* setlocal filetype=dockerfile
autocmd BufRead,BufNewFile *dockerfile* setlocal filetype=dockerfile
autocmd BufRead,BufNewFile *.spec.in setlocal filetype=spec
autocmd BufRead,BufNewFile *.functions setlocal filetype=sh
autocmd BufRead,BufNewFile *.sls setlocal filetype=sls
autocmd BufRead,BufNewFile *.jinja setlocal filetype=jinja

"" General section
set nowrap        " don't wrap lines
set tabstop=4     " a tab is four spaces
set backspace=indent,eol,start
                    " allow backspacing over everything in insert mode
set autoindent    " always set autoindenting on
set copyindent    " copy the previous indentation on autoindenting
set shiftwidth=4  " number of spaces to use for autoindenting
set shiftround    " use multiple of shiftwidth when indenting with '<' and '>'
set showmatch     " set show matching parenthesis
set ignorecase    " ignore case when searching
set smartcase     " ignore case if search pattern is all lowercase,
                    "    case-sensitive otherwise
set smarttab      " insert tabs on the start of a line according to
                    "    shiftwidth, not tabstop
set hlsearch      " highlight search terms
set incsearch     " show search matches as you type

set history=1000         " remember more commands and search history
set undolevels=1000      " use many muchos levels of undo
set wildignore=*.swp,*.bak,*.pyc,*.class
set visualbell           " don't beep
set noerrorbells         " don't beep

" FX button mapping
set pastetoggle=<F2> " Toggle PASTE mode

" Remove trailing whitespaces
:nnoremap <silent> <F5> :let _s=@/ <Bar> :%s/\s\+$//e <Bar> :let @/=_s <Bar> :nohl <Bar> :unlet _s <CR>

set hidden " Allow leaving a buffer with unsaved changes

set mouse=a " Use mouse in all mode

set guioptions-=T  "remove toolbar

" Completion in menu bar
set wildmode=longest,list,full
set wildmenu

"" Remap
" <leader>sv sources .vimrc
nnoremap <leader>sv :source $MYVIMRC<CR>:runtime! plugin/settings/*<CR>:redraw<CR>:echo $MYVIMRC 'reloaded'<CR>

" Buffer navigation
nnoremap <silent> <leader><Left> <esc>:bp<CR>
nnoremap <silent> <leader><Right> <esc>:bn<CR>

" Split navigations
nnoremap <silent> <C-Right> <c-w>l
nnoremap <silent> <C-Left> <c-w>h
nnoremap <silent> <C-Up> <c-w>k
nnoremap <silent> <C-Down> <c-w>j

" Copy to system clipboard Visual selection
" vnoremap <silent> <C-c> y/<C-R>"*<CR><esc>

"" Folding
" Enable folding
set foldmethod=indent
set foldlevel=99

" See the docstring when folding
let g:SimpylFold_docstring_preview=1

" Autocompletion stuff
let g:ycm_autoclose_preview_window_after_completion=1
map <leader>g  :YcmCompleter GoToDefinitionElseDeclaration<CR>

" Colors set
if has('gui_running')
  set background=dark
  colorscheme solarized
  set guifont=DejaVu\ Sans\ Mono\ for\ Powerline\ 9
else
  "colors theme configuration (console)
  if &term == "linux" || &term == "screen.linux"
    set t_Co=8
    colorscheme zenburn
  else
    set t_Co=256
    colorscheme zenburn
  endif
endif

"" NerdTree
let NERDTreeIgnore=['\.pyc$', '\~$'] "ignore files in NERDTree
map <C-n> :NERDTreeToggle<CR>
" Close vim if only NerdTree is open
autocmd bufenter * if (winnr("$") == 1 && exists("b:NERDTree") && b:NERDTree.isTabTree()) | q | endif

" Always enable status bar
set laststatus=2

" Use powerline font symbols
let g:airline_powerline_fonts = 1
if !exists('g:airline_symbols')
  let g:airline_symbols = {}
endif
let g:airline_symbols.space = "\ua0"

" Enable the list of buffers
let g:airline#extensions#tabline#enabled = 1

" Session
let g:session_autosave = "prompt"
let g:session_autosaveperiodic = 60
let g:session_persist_colors = 0
let g:session_persist_font = 0
`
As long as I keep the vim-airline plugin, if I edit a file, :Gwrite, :Gcommit, I end up with a window with the .git/index content like this
`DIRCCX/
Ò¾ÖX/
Ò¾Ö*Ó¤èèd^KOOõ!IÞéÚ8føx|¨.bashrcX/
Ò¾ÖX/
Ò¾Ö*Ô¤èèËâ,ßª7P@å	dáe¥ÒÁza[Á
.gitconfigX/
Ò¾ÖX/
Ò¾Ö*Õ¤èèúóLÇE¤LË®þVEãM$àÊCP
.gitignoreX/
Ò¾ÖX/
Ò¾Ö*Þ¤èèEÖ-IÁ"²ýõú<Ðð1µ.gitignore_globalX/
Ò¾ÖX/
Ò¾Ö*ç¤èèt|Ü2¨s¸Ñx~âï
I¨³âï.gitmodulesX/ÇÕX/ÇÕ*¤Ò¤èè	y±Z¦Âm(np5×`ÊÛnbÓ.vim/.gitignoreX/ÇÕX/ÇÕ*¤Ó¤èè7wq¥+UÉ~{ñåy
×@°].vim/.netrwhistX/ÇÕX/ÇÕ*¤Õ¤èèö:gÖr¹í'KÖ¢Ûø.vim/ftplugin/c.vimX/ÇÕX/ÇÕ*¤Ö¤èèæâ²ÑÖCK)®wZØÂäS.vim/ftplugin/css.vimX/ÇÕX/ÇÕ*¤×¤èèng=¶&Ù·jpÇ%GÆ.vim/ftplugin/go.vimX/ÇÕX/ÇÕ*¤Ø¤èèÎN!µ.äD?T½Ã X¢ÌºC´.vim/ftplugin/html.vimX/ÇÕX/ÇÕ*¤Ù¤èèæâ²ÑÖCK)®wZØÂäS.vim/ftplugin/js.vimX/ÇÕX/ÇÕ*¤Ú¤èè8¨Dùg÷:V²Våùròû¬Cmù.vim/ftplugin/json.vimX/ÇÕX/ÇÕ*¤Û¤èèÎÖhÉ°Æ ïßÜv¬3ÄA:.vim/ftplugin/make.vimX/ÇÕX/ÇÕ*¤Ü¤èèæâ²ÑÖCK)®wZØÂäS.vim/ftplugin/mako.vimX/ÇÕX/ÇÕ*¤Ý¤èèY¢~ÖÍÍàB :Ù0 .vim/ftplugin/md.vimX2É ?sX2É ?s*¤Þ¤èèG·qþ¾FU¸ûÚ¿ R.vim/ftplugin/python.vimX/ÇÕX/ÇÕ*¤ß¤èèâñ¶»J5Gl2ÚÅ¯lýÕ,ºÁ.vim/ftplugin/rst.vimX/ÇÕX/ÇÕ*¤à¤èèæâ²ÑÖCK)®wZØÂäS.vim/ftplugin/rust.vimX/ÇÕX/ÇÕ*¤á¤èèU~Ê
äÔ%þ,^|ÌGªÖf.vim/ftplugin/sh.vimX/ÇÕX/ÇÕ*¤â¤èè8¨Dùg÷:V²Våùròû¬Cmù.vim/ftplugin/spec.vimX/ÇÕX/ÇÕ*¤ã¤èèæâ²ÑÖCK)®wZØÂäS.vim/ftplugin/vala.vimX/ÇÕX/ÇÕ*¤å¤èè	ãë0;-Y ®<Iqmú.vim/plugin/bclose.vimX4Ì»ùgÊX4Ì»¼^+01­¤èè)JY:ðÿüô&É¹ÛP@¬.vimrcX/ÇÕX/ÇÕ*ê¤èè:hG3û»óõÎì½ÍYwZÙÂLICENSEX4Êw6X4Ê:0/+¤èèLûìÖ!ëÀáYzì+÷Ä	README.mdX/ÇÕX/ÇÕ*ñ¤èèìº>9â&kVaíbEÚDmØaliasesX/ÇÕX/ÇÕ*¤è¤èèp7p¦Sâÿ1ÌÇpÊ»backup/exclude/commonX/ÇÕX/ÇÕ*¤é¤èè
ÞÑrÎzù÷\;BÜTþöÙbackup/exclude/scalityX/ÇÕX/ÇÕ*¤í¤èè²çªÌ®ºHìÇtDü¦´3~Â2lib/Linux/default/00-core.shX/ÇÕX/ÇÕ*¤î¤èèqßêf (cL\Z¡´röÇTRlib/Linux/default/01-ssh.shX/ÇÕX/ÇÕ*¤ï¤èè
íùkYy9Ö3¤b1ýåÊKzü&(lib/Linux/default/10-git_aware_prompt.shX/ÇÕX/ÇÕ*¤ð¤èèkï½ÅºjýVä
 Eh(·Ñlib/Linux/default/90-docker.shX/ÇÕX/ÇÕ*¤ñ¤èè³+ó©Äé;×Äb1Ú¥ÇôíûLlib/Linux/default/90-git.shX/ÇÕX/ÇÕ*¤ò¤èèâ,ï¡ëU]¡gú`Rlib/Linux/default/90-go.shX/ÇÕX/ÇÕ*¤ó¤èèÃÀÈ>.oxE`ÁðØ2¶Öºx§lib/Linux/default/90-svn.shX/ÇÕX/ÇÕ*¤ô¤èèµjÓÊOÕçìÍu»F3Âq`\vlib/Linux/default/90-venv.shX/ÇÕX/ÇÕ*¤õ¤èè8óÑùMéÚå~È¡~&PÖlib/Linux/default/90-vim.shX/ÇÕX/ÇÕ*¤ö¤èèo÷kÖ3fKCK¤uÜ&+Plib/Linux/default/99-prompt.shX/ÇÕX/ÇÕ*¤øíèè+µt»Fúò9x÷0íÓQ»\usaltstack/configure.shX/1åé\ÒX/1åé\Ò*¤FíèèHSäõõõa³ã¦!ýz7BÀsaltstack/install.shX/1þ)'Q_X/1þ)'Q_*¦é¤èè=aéÐ_pf!¯¤T
Ê¸× saltstack/salt/common/Debian.slsX/1þ)'Q_X/1þ)'Q_*¦Ó¤èèsýÊUONÈü)hÒ6hÛ}¯3saltstack/salt/common/init.slsX4aL/VnX4aLòM`+¦¤èèª­ïqs»øDv-2Pý.i!saltstack/salt/desktop/Ubuntu.slsX/2b^b;X/2b^b;*¦Ñ¤èè´2¸ÄBÝXq6ÒR3Ï¢saltstack/salt/desktop/init.slsX/ÇÕX/ÇÕ*¦y¤èè0C(¢'[_]vSôÛ	?;ó%saltstack/salt/dev/git/github_com.slsX4.aX4.$x*¦­¤èè¦ö·6ÓçcN >´=óHÒD6'saltstack/salt/dev/git/init.slsX/ÇÕX/ÇÕ*¦|¤èè«æßC÷¢¥÷±k<?°m"saltstack/salt/dev/golang/init.slsX/LÐÔX/LÐÔ*¦}¤èèDü¤I­Æ;xy¶G\I¥»æsaltstack/salt/dev/map.jinjaX4X4*¦¤èèúqø;À-<ïL®Fò©%Oó¼@À"saltstack/salt/dev/python/init.slsX/LÐÔX/LÐÔ*¦¤èè^MùVd&oÕ)=êÁêu¿¨¡PÅ¸!saltstack/salt/dev/shell/init.slsX/LÐÔX/LÐÔ*¦¤èèP`.VàçAí9T`j*ý¿W%@ÖX saltstack/salt/docker/Ubuntu.slsX/LÐÔX/LÐÔ*¦¤èèºÛ¸»]«2gK*?^¬	Ü[,saltstack/salt/docker/files/daemon.json.tmplX/LÐÔX/LÐÔ*¦¤èèÐYr÷]ÙÅ>R±[_J]Ö6zåsaltstack/salt/docker/init.slsX/LÐÔX/LÐÔ*¦¤èè4O}kOm3OaÈ£Ú§õÀsaltstack/salt/editors/init.slsX/LÐÔX/LÐÔ*¦¤èèWO<´Idû«Z?Ó§L#KÊ·ðû)saltstack/salt/editors/vim/completion.slsX/LÐÔX/LÐÔ*¦¤èèl®ïtÐI¬è~=Þ°xÝ/Ú¾¯'saltstack/salt/editors/vim/dotfiles.slsX2ÈÍ!¿X2ÈÍ!¿*¦¤èèu098`Âzd+ÇÃ+AzÕ³ò#saltstack/salt/editors/vim/init.slsX49X49*¦ª¤èè9ñ-#ÎN.º
Rg§°Ò#%saltstack/salt/editors/vim/vundle.slsX/LÐÔX/LÐÔ*¦¤èèç¯õ;£Ü )gcÿ0R!âXsaltstack/salt/fonts/init.slsX4ÃóÑ[JX4ÃóÑ[J0¦¯¤èè"­\	ìe´üöÏJ«½saltstack/salt/init/init.slsX4Åg X4ÅgB0¦¸¤èè@¶æâLËõx¿4Zß]ßqÓ$saltstack/salt/init/systemd/init.slsX4Ç÷.Ð`X4Ç÷.0¦¤èèµÄi.49úiÅàs7~U¿(saltstack/salt/init/systemd/journald.slsX4U X4U *¤"¤èè.¥B~@ÑQe>!6	gósaltstack/salt/lxc/init.slsX/LÐÔX/LÐÔ*¤ú¤èèXa6MÃð×]/[ÊïXVsaltstack/salt/top.slsX4<&¼½X4<&¼½*¦¤èè($ÏO/vù¿ëùä£ð¦ÕW²saltstack/salt/user/env.slsX/LÐÔX/LÐÔ*¦¤èèüCUOºóÊk:'¾£¾lsaltstack/salt/user/info.slsTREE667 4
¬ùÿÓn:q±H_à/lib10 1
'ÄÎ/üî
§É£Î¾ÞLinux10 1
N0ÃÒ%5gçlÉ¥¨_jödefault10 0
3Ã?ºÄ`åfÏJhß\&µ.vim18 2`

The title of the window being "<PATH to git repo>/.git/index" [Preview]
I can just close it but it's rather annoying

As long as I remove vim-airline plugin, this does not happen anymore

**Comment by darkalia** [comment]
Found it. This regression was introduced by https://github.com/vim-airline/vim-airline/commit/3899f848a751b2f8e6d45fa1fe3f723491999f6f
If you put your vim-airline to 02ecb8631d4c4252e11fff2ed9fc8d48d91577a1, it works fine
I'll put that to vim-airline repository also

**Comment by zoni** [comment]
According to vim-airline maintainers, this is an issue in fugitive rather than vim-airline. https://github.com/vim-airline/vim-airline/issues/1332#issuecomment-263123809 has a detailed explanation of what is happening here.

**Comment by tpope** [comment]
If I understand correctly, that ought to do it.

**Comment by zoni** [comment]
Looks like we're not quite there yet. Running with https://github.com/tpope/vim-fugitive/commit/c5c1bd66d897501848312a97b408ca5255251a9a, issuing `:Gcommit` now errors with:

```
Error detected while processing function <SNR>45_Commit:
line   25:
E46: Cannot change read-only variable "shell_error"
Press ENTER or type command to continue
```

**Comment by tpope** [comment]
Crap, try now.

**Comment by zoni** [comment]
Success! :smile: I can confirm that fixed it (that is both the regression from c5c1bd66d897501848312a97b408ca5255251a9a as well as the original issue itself).

Thanks!
